File: math_and_geometry/202_happy_number/202.cpp


class Solution {
public:
    
    int get_next(int num) {
        int next_num = 0;
        while (num != 0) {
            next_num += ((num % 10) * (num % 10));
            num = (int)(num / 10);
        }
        return next_num;
    }
    
    bool isHappy(int n) {
        int slow = n;
        int fast = get_next(slow);
        while (fast != 1 && slow != fast) {
            slow = get_next(slow);
            fast = get_next(get_next(fast));
        }
        return fast == 1;
    }
};

File: math_and_geometry/202_happy_number/202.rs


impl Solution {
    pub fn is_happy(n: i32) -> bool {
        
        fn get_next(num: i32) -> i32 {
            let mut next_num: i32 = 0;
            let mut num = num;
            while num != 0 {
                next_num += (num % 10).pow(2);
                num = (num / 10);
            }
            return next_num;
        }
        
        let mut slow: i32 = n;
        let mut fast: i32 = get_next(slow);
        while fast != 1 && slow != fast {
            slow = get_next(slow);
            fast = get_next(get_next(fast));
        }
        return fast == 1;
    }
}